File: python/Bejeweled_Chains.py
def checkVertical(arr, row_index, col_index, max_rows, VALID_COUNT):
    index = row_index
    num = arr[row_index][col_index]
    count = 0

    if row_index > 0:
        start_row_index = row_index-1
    elif row_index == 0:
        start_row_index = 0
    
    if row_index == 0:
        while index <= max_rows-1:
            if num == arr[index][col_index]:
                count += 1
            else:
                break
            index += 1
    elif start_row_index >= 0 and num != arr[start_row_index][col_index]:
        while index <= max_rows-1:
            if num == arr[index][col_index]:
                count += 1
            else:
                break
            index += 1

    if count >= VALID_COUNT:
        return True
    return False

def checkHorizontal(arr, row_index, col_index, max_cols, VALID_COUNT):
    index = col_index
    num = arr[row_index][col_index]
    count = 0

    if col_index > 0:
        start_col_index = col_index-1
    elif col_index == 0:
        start_col_index = 0

    if col_index == 0:
        while index <= max_cols-1:
            if num == arr[row_index][index]:
                count += 1
            else:
                break
            index += 1
    elif start_col_index >= 0 and num != arr[row_index][start_col_index]:
        while index <= max_cols-1:
            if num == arr[row_index][index]:
                count += 1
            else:
                break
            index += 1

    if count >= VALID_COUNT:
        return True
    return False

# Time Complexity: O(N*M*(N-1+M-1)). Given N is the number of elements in a row and M is the number of elements in a column.
# Space Complexity: O(1). No auxiliary memory used.
def count_chains(arr):
    max_rows = len(arr)
    max_cols = len(arr[0])
    VALID_COUNT = 3

    count = 0
    for row_index in range(len(arr)):
        for col_index in range(len(arr[row_index])):
            valid_horizontal = checkHorizontal(arr, row_index, col_index, max_cols, VALID_COUNT)
            valid_vertical = checkVertical(arr, row_index, col_index, max_rows, VALID_COUNT)
            if valid_horizontal and not valid_vertical:
                count += 1
            elif valid_vertical and not valid_horizontal:
                count += 1
            elif valid_vertical and valid_horizontal:
                count += 2

    return count

assert count_chains([[0,3,3],[0,0,0],[0,1,4]]) == 2, "Wrong answer for [[0,3,3],[0,0,0],[0,1,4]]"
assert count_chains([[3,3,3,4,1,2],[0,0,3,2,4,4],[4,0,0,2,2,1],[1,3,4,2,4,0],[0,1,0,2,0,1],[2,2,2,4,1,3]]) == 3, "Wrong answer for [[3,3,3,4,1,2],[0,0,3,2,4,4],[4,0,0,2,2,1],[1,3,4,2,4,0],[0,1,0,2,0,1],[2,2,2,4,1,3]]"
assert count_chains([[3,3,3,3,3],[0,0,1,2,1],[0,1,4,3,2]]) == 1, "Wrong answer for [[3,3,3,3,3],[0,0,1,2,1],[0,1,4,3,2]]"

File: go/Bejeweled_Chains.go
package main

func checkVertical(arr [][]int, row_index int, col_index int, max_rows int, VALIDCOUNT int) bool {
	index := row_index
    num := arr[row_index][col_index]
    count := 0
	start_row_index := 0

    if row_index > 0 {
		start_row_index = row_index-1
	} else if row_index == 0 {
		start_row_index = 0
	}

    if row_index == 0 || (start_row_index >= 0 && num != arr[start_row_index][col_index]) {
		for {
			if index > max_rows-1 {
				break
			}
			if num == arr[index][col_index] {
				count += 1
			} else {
				break
			}
			index = index + 1
		}
	}
        

    if count >= VALIDCOUNT {
		return true
	}
    return false
}

func checkHorizontal(arr [][]int, row_index int, col_index int, max_cols int, VALIDCOUNT int) bool {
	index := col_index
    num := arr[row_index][col_index]
    count := 0
	start_col_index := 0

    if col_index > 0 {
		start_col_index = col_index-1
	} else if col_index == 0 {
		start_col_index = 0
	}

    if col_index == 0 || (start_col_index >= 0 && num != arr[row_index][start_col_index]) {
		for {
			if index > max_cols-1 {
				break
			}
			if num == arr[row_index][index] {
				count += 1
			} else {
				break
			}
			index = index + 1
		}
	}
        

    if count >= VALIDCOUNT {
		return true
	}
    return false
}

// Time Complexity: O(N*M*(N-1+M-1)). Given N is the number of rows of the array and M is the number of elements in the row.
// Space Complexity: O(1)
func count_chains(arr [][]int) int {
	maxRows := len(arr)
	maxCols := len(arr[0])
	VALIDCOUNT := 3
	validHorizontal := false
	validVertical := false

	count := 0
	for rowIndex, _ := range arr {
		for colIndex, _ := range arr[rowIndex] {
			validHorizontal = checkHorizontal(arr, rowIndex, colIndex, maxCols, VALIDCOUNT)
            validVertical = checkVertical(arr, rowIndex, colIndex, maxRows, VALIDCOUNT)
            if validHorizontal == true && validVertical == false{
				count = count + 1
			} else if validVertical == true && validHorizontal == false {
				count = count + 1
			} else if validVertical == true && validHorizontal == true {
				count = count + 2
			}
		}
	}

	return count
}

func main() {
	test01 := [][]int{
		{0,3,3,},
		{0,0,0,},
		{0,1,4,},
	}
	ans01 := count_chains(test01)
	if ans01 != 2 {
		panic("Wrong answer for [[0,3,3],[0,0,0],[0,1,4]]")
	}

	test02 := [][]int{
		{3,3,3,4,1,2,},
		{0,0,3,2,4,4,},
		{4,0,0,2,2,1,},
		{1,3,4,2,4,0,},
		{0,1,0,2,0,1,},
		{2,2,2,4,1,3,},
	}
	ans02 := count_chains(test02)
	if ans02 != 3 {
		panic("Wrong answer for [[3,3,3,4,1,2],[0,0,3,2,4,4],[4,0,0,2,2,1],[1,3,4,2,4,0],[0,1,0,2,0,1],[2,2,2,4,1,3]]")
	}

	test03 := [][]int{
		{3,3,3,3,3,},
		{0,0,1,2,1,},
		{0,1,4,3,2,},
	}
	ans03 := count_chains(test03)
	if ans03 != 1 {
		panic("Wrong answer for [[3,3,3,3,3],[0,0,1,2,1],[0,1,4,3,2]]")
	}
}